src/mlpack/core/tree/greedy_single_tree_traverser_impl.hpp

```diff
@@ -24,26 +24,48 @@ template<typename TreeType, typename RuleType>
 GreedySingleTreeTraverser<TreeType, RuleType>::GreedySingleTreeTraverser(
     RuleType& rule) :
     rule(rule),
-    numPrunes(0)
+    numPrunes(0),
+    k(0)
 { /* Nothing to do. */ }
 
 template<typename TreeType, typename RuleType>
 void GreedySingleTreeTraverser<TreeType, RuleType>::Traverse(
     const size_t queryIndex,
     TreeType& referenceNode)
 {
-  // Run the base case as necessary for all the points in the reference node.
-  for (size_t i = 0; i < referenceNode.NumPoints(); ++i)
-    rule.BaseCase(queryIndex, referenceNode.Point(i));
+  if(referenceNode.IsLeaf())
+  {
+    // Run the base case as necessary for all the points in the reference node.
+    for (size_t i = 0; i < referenceNode.NumPoints(); ++i)
+      rule.BaseCase(queryIndex, referenceNode.Point(i));
+    return;
+  }
+
+  size_t bestChild = rule.GetBestChild(queryIndex, referenceNode);
+  size_t numDescendants;
 
-  if (!referenceNode.IsLeaf())
+  // Check that referencenode is not a leaf node while calculating number of
+  // descendants of it's best child.
+  if(!referenceNode.IsLeaf())
+    numDescendants = referenceNode.Child(bestChild).NumDescendants();
```

Comment:
We need to calculate `numDescendant` with `!referenceNode.IsLeaf()` condition imposed, otherwise on asking for child of leaf node it will give us segmentation fault.